### ISSUE TITLE
read-package-tree assumes the path it's passed exists

### ISSUE THREAD

**Issue by othiym23** [npm/npm#8020]
This is the proximate cause for `node test/run.js` exploding. I'm still puzzling out what the root cause is, and how to fix this, but here's a snapshot of what I've figured out so far:

* `test/run.js` [sets `global` to true](https://github.com/npm/npm/blob/master/test/run.js#L46) via the environment
* the `globalDir` accessor returns a value of `prefix` + `/lib` as the [location used by install](https://github.com/npm/npm/blob/master/test/run.js#L46). At this point, only `prefix` and `prefix` + `/etc` exist.
* `readLocalPackageData()` [calls out to `readPackageTree()`] with `globalDir` as the value for `where`, which doesn't yet exist, and `readPackageTree()` blows up

I suspect, but am not yet entirely sure, that instead of using `iferr` there, we should be testing for `ENOENT` and calling some kind of generic utility function that creates the base directory if it doesn't yet exist. Either that, or there should be a previous step that ensures that the baseline conditions for a successful install are in place. It's rare that we're going to be using npm to install to an entirely new (global) prefix, but it's definitely something that should be supported.

**Comment by iarna** [comment]
So these are the errors I get:
```
🕠  rebecca@Caldina:~/code/npm$ [multi-stage *] node test/run.js
# testing in /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741
# global prefix = /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root

+node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
$$$$$$ cd /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root; PATH=/Users/rebecca/.nvm/v0.10.38/bin:/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/bin:/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/node_modules/.bin:/Users/rebecca/.nvm/v0.10.38/bin:/opt/local/libexec/gnubin:/opt/local/bin:/opt/local/sbin:/Users/rebecca/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin /Users/rebecca/.nvm/v0.10.38/bin/node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
$$$$$$ after command /Users/rebecca/.nvm/v0.10.38/bin/node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm" /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root
 2> npm ERR! Darwin 14.3.0
 2> npm ERR! argv "/Users/rebecca/.nvm/v0.10.38/bin/node" "/Users/rebecca/code/npm/bin/npm-cli.js" "install" "/Users/rebecca/code/npm"
 2> npm ERR! node v0.10.38
 2> npm ERR! npm  v3.0.0-pre.6
 2> npm ERR! path /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/lib
 2> npm ERR! code ENOENT
 2> npm ERR! errno 34
 2>
 2> npm ERR! enoent ENOENT, lstat '/private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/lib'
 2> npm ERR! enoent This is most likely not a problem with npm itself
 2> npm ERR! enoent and is related to npm not being able to find a file.
 2> npm ERR! enoent
 2>
 2> npm ERR! Please include the following file with any support request:
 2> npm ERR!     /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/npm-debug.log
 2>
not ok 1 node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
1..1

/Users/rebecca/code/npm/test/run.js:192
  if (er) throw er
                ^
Error: failed node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
    at /Users/rebecca/code/npm/test/run.js:104:10
    at ChildProcess.exithandler (child_process.js:662:7)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Socket.<anonymous> (child_process.js:979:11)
    at Socket.emit (events.js:95:17)
    at Pipe.close (net.js:466:12)
🕠  rebecca@Caldina:~/code/npm$ [multi-stage *] ls -l /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/lib
ls: cannot access /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root/lib: No such file or directory
🕠  rebecca@Caldina:~/code/npm$ [multi-stage *] ls -l /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49741/root
total 4
drwxr-xr-x 2 rebecca staff   68 Apr 21 05:15 etc
-rw-r--r-- 1 rebecca staff 1171 Apr 21 05:15 npm-debug.log
```

**Comment by iarna** [comment]
So I tried putting in a blind mkdirp on the folder we're to install into:
```
🕠rebecca@Caldina:~/code/npm$ [multi-stage *] node test/run.js
# testing in /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866
# global prefix = /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root

+node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
$$$$$$ cd /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root; PATH=/Users/rebecca/.nvm/v0.10.38/bin:/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root/bin:/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root/node_modules/.bin:/Users/rebecca/.nvm/v0.10.38/bin:/opt/local/libexec/gnubin:/opt/local/bin:/opt/local/sbin:/Users/rebecca/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin /Users/rebecca/.nvm/v0.10.38/bin/node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
$$$$$$ after command /Users/rebecca/.nvm/v0.10.38/bin/node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm" /var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root
 1> > npm@3.0.0-pre.6 prepublish /Users/rebecca/code/npm
 1> > node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 doc
 2> sh: rimraf: command not found
 2>
 2> npm ERR! addLocal Could not install /Users/rebecca/code/npm
 2> npm ERR! Darwin 14.3.0
 2> npm ERR! argv "/Users/rebecca/.nvm/v0.10.38/bin/node" "/Users/rebecca/code/npm/bin/npm-cli.js" "install" "/Users/rebecca/code/npm"
 2> npm ERR! node v0.10.38
 2> npm ERR! npm  v3.0.0-pre.6
 2> npm ERR! file sh
 2> npm ERR! code ELIFECYCLE
 2> npm ERR! errno ENOENT
 2> npm ERR! syscall spawn
 2> npm ERR! npm@3.0.0-pre.6 prepublish: `node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 doc`
 2> npm ERR! spawn ENOENT
 2> npm ERR!
 2> npm ERR! Failed at the npm@3.0.0-pre.6 prepublish script 'node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 doc'.
 2> npm ERR! This is most likely a problem with the npm package,
 2> npm ERR! not with npm itself.
 2> npm ERR! Tell the author that this fails on your system:
 2> npm ERR!     node bin/npm-cli.js prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 doc
 2> npm ERR! You can get their info via:
 2> npm ERR!     npm owner ls npm
 2> npm ERR! There is likely additional logging output above.
 2>
 2> npm ERR! Please include the following file with any support request:
 2> npm ERR!     /private/var/folders/56/cm7r6x9n3h78yb6k6md53k940000gn/T/npm-test-49866/root/npm-debug.log
 2>
not ok 1 node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
1..1

/Users/rebecca/code/npm/test/run.js:192
  if (er) throw er
                ^
Error: failed node "/Users/rebecca/code/npm/bin/npm-cli.js" install "/Users/rebecca/code/npm"
    at /Users/rebecca/code/npm/test/run.js:104:10
    at ChildProcess.exithandler (child_process.js:662:7)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Socket.<anonymous> (child_process.js:979:11)
    at Socket.emit (events.js:95:17)
    at Pipe.close (net.js:466:12)
```

**Comment by thefourtheye** [comment]
`baseline conditions for a successful install` Can we list them down here?

**Comment by othiym23** [comment]
@thefourtheye Sure, once I figure them out. There are a number of paths that need to exist before the installer runs, like the `prefix` and some of its subdirectories, depending on how npm is invoked and configured. I think the patch you and @iarna tried should work, it's just that first test step is very time-consuming because it builds the docs as part of installing npm, which is itself very time-consuming.

**Comment by thefourtheye** [comment]
@othiym23 @iarna Please check if #8028 looks okay.

**Comment by othiym23** [comment]
This is fixed by #8028. Thanks, @thefourtheye and @iarna!
